File: services/hierarchy/lib/domain/sensor.ml
type meter_type = Counter | Gauge

type t = {
  id : Sensor_id.t;
  created : Ptime.t;
  daq_id : string;
  (* Pipe-separated list of ancestor ids from root down to and including
     self. The trailing segment is the sensor id (S#<n>); the segment
     before it is the direct parent node-id (HN<n>#<n>). In storage this
     string is also the gsi1sk attribute. *)
  path : string;
  purpose : string;
  meter_type : meter_type;
  unit : string option;
  formula : Formula.t;
  (* Resample interval in minutes. Typical values: 5, 15, 60. When set,
     must be > 0; the downstream timeseries store uses it to know how to
     resample raw readings onto a fixed time grid. None = no resampling
     configured for this sensor. *)
  resample_minutes : int option;
}

let path_sep = "|"

let child_path ~parent_path ~sensor_id_str =
  parent_path ^ path_sep ^ sensor_id_str

(* Walk the path from the back and return the last HN<n>#<uuid> segment
   as a Node_id. Sensors must always have a node parent, so this raises
   if the path is malformed. *)
let parent_id (t : t) : Node_id.t =
  let parts = String.split_on_char '|' t.path |> List.filter (fun s -> s <> "") in
  let rec last_hn = function
    | [] -> None
    | seg :: rest ->
        (match Node_id.of_string seg with
         | Ok id -> (match last_hn rest with Some x -> Some x | None -> Some id)
         | Error _ -> last_hn rest)
  in
  match last_hn parts with
  | Some id -> id
  | None ->
      failwith (Printf.sprintf "Sensor.parent_id: no node segment in path %S" t.path)

let meter_type_to_string = function
  | Counter -> "counter"
  | Gauge   -> "gauge"

let meter_type_of_string = function
  | "counter" -> Ok Counter
  | "gauge"   -> Ok Gauge
  | other     -> Error (Printf.sprintf "unknown meter type %S" other)

File: crates/model/src/domain/sensor.rs
use chrono::{DateTime, Utc};
use typed_builder::TypedBuilder;

use crate::domain::formula::Formula;
use crate::domain::ids::{NodeId, SensorId};
use crate::domain::node::PATH_SEP;
use crate::domain::values::{ReadingKind, EnergyType};

// ---------------------------------------------------------------------------
// Sensor
// ---------------------------------------------------------------------------

/// A sensor attached to a hierarchy node.
///
/// `path` is a pipe-separated list of ancestor ids from root down to and
/// including self (the `gsi1sk` attribute in DynamoDB).
#[derive(Clone, Debug, PartialEq, TypedBuilder)]
pub struct Sensor {
    pub id: SensorId,
    #[builder(default = chrono::Utc::now())]
    pub created: DateTime<Utc>,
    pub daq_id: String,
    pub path: String,
    pub energy_type: EnergyType,
    pub reading_kind: ReadingKind,
    #[builder(default)]
    pub unit: Option<String>,
    #[builder(default = Formula::Identity)]
    pub formula: Formula,
    #[builder(default)]
    pub resample_minutes: Option<i32>,
}

// ---------------------------------------------------------------------------
// Accessors
// ---------------------------------------------------------------------------

impl Sensor {
    /// Walk the path from the back and return the last `HN<n>#<id>` segment
    /// as a `NodeId`.  Sensors must always have a node parent.
    ///
    /// Panics if the path contains no valid node-id segment.
    pub fn parent_id(&self) -> NodeId {
        let parts: Vec<&str> = self
            .path
            .split(PATH_SEP)
            .filter(|s| !s.is_empty())
            .collect();

        // Walk forward, keeping the last successfully parsed NodeId.
        let mut last: Option<NodeId> = None;
        for seg in &parts {
            if let Ok(nid) = NodeId::parse(seg) {
                last = Some(nid);
            }
        }
        last.unwrap_or_else(|| {
            panic!("Sensor.parent_id: no node segment in path {:?}", self.path)
        })
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::ids::Level;
    use crate::domain::node::child_path;
    use chrono::DateTime;

    fn ptime_of(s: &str) -> DateTime<Utc> {
        DateTime::parse_from_rfc3339(s)
            .unwrap()
            .with_timezone(&Utc)
    }

    fn make_sample() -> Sensor {
        let parent = NodeId::make(Level::Hn5, 10042);
        let id = SensorId::make(20001);
        let parent_path = format!(
            "{}|{}",
            NodeId::root().to_string(),
            parent.to_string()
        );
        let path = child_path(&parent_path, &id.to_string());
        Sensor::builder()
            .id(id)
            .created(ptime_of("2026-04-18T10:00:00Z"))
            .daq_id("daq:adeunis_pu_v1:123:0018b210000191c7:counter_a".to_owned())
            .path(path)
            .energy_type(EnergyType::Electricity)
            .reading_kind(ReadingKind::Counter)
            .unit(Some("kWh".to_owned()))
            .formula(Formula::Identity)
            .resample_minutes(Some(15))
            .build()
    }

    #[test]
    fn fields_preserved() {
        let s = make_sample();
        assert_eq!(s.energy_type, EnergyType::Electricity);
        assert_eq!(s.daq_id, "daq:adeunis_pu_v1:123:0018b210000191c7:counter_a");
        assert!(matches!(s.reading_kind, ReadingKind::Counter));
        assert_eq!(s.unit, Some("kWh".to_owned()));
    }

    #[test]
    fn parent_id_extracts_last_node_segment() {
        let s = make_sample();
        let expected = NodeId::make(Level::Hn5, 10042);
        assert_eq!(
            s.parent_id().to_string(),
            expected.to_string(),
            "parent id from path"
        );
    }

    /// child_path joins with `|`.
    #[test]
    fn child_path_join() {
        let p = child_path("HN0#root|HN5#10042", "S#20001");
        assert_eq!(p, "HN0#root|HN5#10042|S#20001");
    }

    /// parent_id panics for a path with no node segments.
    #[test]
    #[should_panic(expected = "Sensor.parent_id")]
    fn parent_id_panics_no_node_segment() {
        let s = Sensor::builder()
            .id(SensorId::make(1))
            .created(ptime_of("2026-01-01T00:00:00Z"))
            .daq_id("x".to_owned())
            .path("S#1".to_owned()) // no HN segment
            .energy_type(EnergyType::Water)
            .reading_kind(ReadingKind::Counter)
            .build();
        let _ = s.parent_id();
    }
}
